File: Python/maximum-score-of-a-good-subarray.py
# Time:  O(n)
# Space: O(n)

class Solution(object):
    def maximumScore(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        result = curr = nums[k]
        left = right = k
        while left-1 >= 0 or right+1 < len(nums):
            # choosing larger one to expand is always better than or equal to choosing smaller one
            if (nums[left-1] if left-1 >= 0 else 0) <= (nums[right+1] if right+1 < len(nums) else 0):
                right += 1
            else:
                left -= 1
            curr = min(curr, nums[left], nums[right])
            result = max(result, curr*(right-left+1))
        return result


# Time:  O(nlogn)
# Space: O(n)
import bisect


class Solution2(object):
    def maximumScore(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        def score(nums, k):
            prefix = [nums[k]]*(k+1)
            for i in reversed(xrange(k)):
                prefix[i] = min(prefix[i+1], nums[i])
            result = right = nums[k]
            for j in xrange(k+1, len(nums)):
                right = min(right, nums[j])
                i = bisect.bisect_left(prefix, right)
                if i >= 0:
                    result = max(result, right*(j-i+1))
            return result

        return max(score(nums, k), score(nums[::-1], len(nums)-1-k))
 

File: C++/maximum-score-of-a-good-subarray.cpp
// Time:  O(n)
// Space: O(1)

class Solution {
public:
    int maximumScore(vector<int>& nums, int k) {
        int result = nums[k], curr = nums[k];
        int left = k, right = k;
        while (left - 1 >= 0 || right + 1 < size(nums)) {
            int l = (left - 1 >= 0 ? nums[left - 1] : 0);
            int r = (right + 1 < size(nums) ? nums[right + 1] : 0);
            if (l < r) { // choosing larger one to expand is always better than or equal to choosing smaller one
                ++right;
            } else {
                --left;
            }
            curr = min({curr, nums[left], nums[right]});
            result = max(result, curr * (right - left + 1));
        }
        return result;
    }
};

// Time:  O(nlogn)
// Space: O(n)
class Solution2 {
public:
    int maximumScore(vector<int>& nums, int k) {
        int result = score(nums, k);
        reverse(begin(nums), end(nums));
        return max(result, score(nums, size(nums) - k - 1));
    }

private:
    int score(const vector<int>& nums, int k) {
        vector<int> prefix(k + 1, nums[k]);
        for (int i = k - 1; i >= 0; --i) {
            prefix[i] = min(prefix[i + 1], nums[i]);
        }
        int result = nums[k], right = nums[k];
        for (int j = k; j < size(nums); ++j) {
            right = min(right, nums[j]);
            const int i = distance(cbegin(prefix), lower_bound(cbegin(prefix), cend(prefix), right));
            if (i != size(prefix)) {
                result = max(result, right * (j - i + 1));
            }
        }
        return result;
    }
};
